src/cargo/sources/registry/remote.rs

```diff
@@ -203,11 +203,15 @@ impl<'cfg> RegistryData for RemoteRegistry<'cfg> {
         self.config.shell().status("Downloading", pkg)?;
 
         let config = self.config()?.unwrap();
-        let mut url = config.dl.to_url()?;
-        url.path_segments_mut().unwrap()
-            .push(pkg.name())
-            .push(&pkg.version().to_string())
-            .push("download");
+        let mut url = config.dl.clone();
+        if !url.contains(CRATE_TEMPLATE) && !url.contains(VERSION_TEMPLATE) {
+            let suffix = format!("/{}/{}/download", CRATE_TEMPLATE, VERSION_TEMPLATE);
```

Comment:
Total nit but this intermediate allocation bothers me, can't we do `write!(url, "/{}/{}/downloads", ...)`?

Reply:
Yep that'd also work.